### ISSUE TITLE
Re-take: Parameters.copy() is broken

### ISSUE THREAD

**Issue by tronsgaard** [lmfit/lmfit-py#298]
As requested by @newville, I am closing #296 and #297 and starting over here:

The following (silly) example illustrates the issue:
```python
from lmfit import Parameters
p1 = Parameters()
p2 = p1.copy()
p2.add('foo')
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.4/dist-packages/lmfit/parameter.py in _getval(self)
    546         try:
--> 547             self.value = max(self.min, min(self._val, self.max))
    548         except(TypeError, ValueError):

TypeError: unorderable types: float() < NoneType()

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
<ipython-input-5-b0051757b071> in <module>()
----> 1 p2.add('foo')

/usr/local/lib/python3.4/dist-packages/lmfit/parameter.py in add(self, name, value, vary, min, max, expr)
    198         else:
    199             self.__setitem__(name, Parameter(value=value, name=name, vary=vary,
--> 200                                              min=min, max=max, expr=expr))
    201 
    202     def add_many(self, *parlist):

/usr/local/lib/python3.4/dist-packages/lmfit/parameter.py in __setitem__(self, key, par)
     83         par.name = key
     84         par._expr_eval = self._asteval
---> 85         self._asteval.symtable[key] = par.value
     86 
     87     def __add__(self, other):

/usr/local/lib/python3.4/dist-packages/lmfit/parameter.py in value(self)
    557     def value(self):
    558         "The numerical value of the Parameter, with bounds applied"
--> 559         return self._getval()
    560 
    561     @value.setter

/usr/local/lib/python3.4/dist-packages/lmfit/parameter.py in _getval(self)
    547             self.value = max(self.min, min(self._val, self.max))
    548         except(TypeError, ValueError):
--> 549             self.value = nan
    550         return self._val
    551 

/usr/local/lib/python3.4/dist-packages/lmfit/parameter.py in value(self, val)
    565         if not hasattr(self, '_expr_eval'):  self._expr_eval = None
    566         if self._expr_eval is not None:
--> 567             self._expr_eval.symtable[self.name] = val
    568 
    569     @property

AttributeError: 'Parameters' object has no attribute 'symtable'
```

Although there might not be the same need for copying parameters after version 0.9, all methods visible to the user should be working as expected or at least raise a more meaningful error. It should also be more clear to the user whether `.copy()` performs a shallow or deep copy.

The error is caused by a change in the definition of `Parameters.__init__()`. When `.copy()` calls `__init__(self)`, `self` is mistakenly interpreted as the `asteval` keyword argument:

```python
## OrderedDict

    def copy(self):
        'od.copy() -> a shallow copy of od'
        return self.__class__(self)
```

```python
## Parameters

    def __init__(self, asteval=None, *args, **kwds):
        super(Parameters, self).__init__(self)
        self._asteval = asteval

        if asteval is None:
            self._asteval = Interpreter()
        self.update(*args, **kwds)
```

This can be confirmed with the following example (continuing from the first example):

```python
p2._asteval.add('foo')
print(p2._asteval)
# Parameters([('foo', <Parameter 'foo', -inf, bounds=[-inf:inf]>)])
```

Rearranging the arguments (`*args, asteval=None, **kwds` instead of `asteval=None, *args, *kwds`) solves the issue in Python3, but it is not valid Python2 code.

As suggested by @newville, a possible fix could be simply replacing `.copy()` with an alias of `.__deepcopy__()`. This is probably the functionality one would expect as a user - however `.copy()` used to return a shallow copy. Another suggestion is to make `.copy()` raise a `NotImplementedError`.

**Comment by newville** [comment]
@tronsgaard   Thanks for starting over.  I can confirm that Parameters.copy() is not working well, and suggest that we explicitly implement a `copy()` method.  

I think it's OK to say that shallow copy of a Parameters instance is not very useful, and that `copy()` should always be a deepcopy with a fresh instance of asteval.    Then again, if pressed, I might say that I always find shallow copies to be confusing and error-prone, so maybe I'm missing a subtle use-case for them.

**Comment by andyfaff** [comment]
Closed via #300
